### ISSUE TITLE
solv_xfopen() on a damaged archive doesn't fail in fread()

### ISSUE THREAD

**Issue by lukash** [openSUSE/libsolv#373]
When uncompressing an (in this case gzip) archive with `solv_xfopen()`, if the data has just been cut off, it doesn't return an error in `fread()`, only in `fclose()`. Only if the data really is garbage, an error from `fread()` is returned.

The problem with this is that the user gets an error that reads "Error closing file 'foo'." Which doesn't make much sense, they should be getting "Error reading file 'foo'." and a more specific error message would be even better, but the FILE interface doesn't lend itself to that... (at least in this case we can say with certainty the problem is a damaged archive)

Create a test file with this script:

```bash
#!/bin/bash

echo "data" > test-file
gzip test-file
head -c-4 test-file.gz > damaged-test-file.gz
# uncomment to get a fread() error
#echo "garbage" >> damaged-test-file.gz
```

And reproduce:
```C
#include <stdio.h>  // libsolv is actually dependent on this being included first(!)
#include <solv/solv_xfopen.h>

#include <errno.h>
#include <string.h>


int main(int argc, char** argv) {
    errno = 0;

    FILE* file = solv_xfopen("damaged-test-file.gz", "r");

    if (!file) {
        printf("Error opening file: %s\n", strerror(errno));
        return 1;
    }

    errno = 0;

    char buffer[1024];
    size_t r = fread(buffer, sizeof(char), sizeof(buffer), file);

    if (r != sizeof(buffer) && ferror(file) != 0) {
        printf("Error reading file: %s\n", strerror(errno));
        return 1;
    }

    errno = 0;

    if (fclose(file) != 0) {
        printf("Error closing file: %s\n", strerror(errno));
        return 1;
    }

    return 0;
}
```

**Comment by mlschroe** [comment]
I agree, but it's actually zlib's gzread that"s to blame. It returns zero if the state is Z_BUF_ERROR, i.e. if there is an unexpected EOF:
```
    /* read len or fewer bytes to buf */
    len = gz_read(state, buf, len);

    /* check for an error */
    if (len == 0 && state->err != Z_OK && state->err != Z_BUF_ERROR)
        return -1;

    /* return the number of bytes read (this is assured to fit in an int) */
    return (int)len;
```

As a workaround we can call gzerror() if we get back zero bytes and return an error for Z_BUF_ERROR.

At some point in time it is probably better to use the low-level functions instead of the gz helpers.

**Comment by mlschroe** [comment]
It's actually documented in the zlib.h header:
```
Note that gzread does not return –1 in the event of an incomplete gzip stream. This error is
deferred until gzclose(), which will return Z_BUF_ERROR if the last gzread ended in the middle
of a gzip stream. Alternatively, gzerror can be used before gzclose to detect this case. 
```

Ok, let's do the gzerror approach.

**Comment by mlschroe** [comment]
Fixed with commit 5013bf78c88188005e9d4950faf2ff91663159b6

**Comment by lukash** [comment]
Thanks!
